File: cookbook/php/duplicateAndAddNewSegment.php
<?php

//Url de base
$urlBase = 'http://v8.mailperformance.com/';

//Ici, renseignez l'email de la cible, la X-Key
$xKey = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
$unicity = 'test@test.com';	//L'adresse mail de la cible a mettre dans le segment ('null' pour ne rien ajouter au segment)
$segmentId = 0123;	//Id du segment a modifier ('null' si le segment est a creer)
$actionId = "000ABC";	//Id de l'action a dupliquer


function getTargetJson($unicity)
{
	//Syntaxe pour les différents types d'informations de la cible :
	$targetString = "name";	//Chaine de caracteres
	$tagetListOfValues = "Mr";	//Liste de valeurs
	$targetEmail = $unicity;	// E-mail
	$tagetPhoneNumber = "0123456789";	// Telephone
	$targetTextZone = "150 caracters max";	//Zone de texte
	$targetNumbers = 123;	//Valeur numérique
	$targetDate = "01/01/2000";	//Date
	$largetListMultipleValues = array("valeur 1", "valeur 2");	//Liste de valeurs multiples
	
	//Creation du tableau en fonction de l'id des champs de la fiche cible : "id-champ" => "valeur de l'information"
	$targetData = array("5398" => $targetString,
			"5399" => $tagetListOfValues,
			"5400" => $targetEmail,
			"5452" => $tagetPhoneNumber,
			"5453" => $targetTextZone,
			"5454" => $targetNumbers,
			"5455" => $targetDate,
			"5456" => $largetListMultipleValues);
	$targetJson = json_encode($targetData);
	
	return ($targetJson);
}

function getSegmentJson()
{
	$segmentType = 'static';	//Code pour creer un segment Statique
	$segmentName = 'Nom du segment';	//Nom du segment
	$segmentDescription = 'Description';	//Description du segment
	$segmentExpiration = '2016-09-23T11:37:49.686Z';	//Date d'expiration du segment
	$segmentIsTest = true;	//Le test est un segment de test : oui = 'true' / non = 'false'	
		
	//Creation du Json du message
	$segmentData = array(
		'type' => $segmentType,
		'name' => $segmentName,
		'description' => $segmentDescription,
		'expiration' => $segmentExpiration,
		'isTest' => $segmentIsTest);
	
	//On affiche le message
	$segmentJson = json_encode($segmentData);
	echo $segmentJson . "\n";
	
	return ($segmentJson);
}

function getNewActionDuplicateJson($actionDuplicateJson, $segmentId)
{
	$actionDuplicateJson = json_decode($actionDuplicateJson, true);
	
	//Modification de la nouvelle action
	$actionDuplicateJson['scheduler'] = array(
		"type" => "asap",	//Envoie : immédiat = 'asap' / Date = 'scheduled'
		//"type" => "2015-07-27T08:15:00Z",	//Si type = 'scheduled' sinon a enlever
		"segments" => array(
			"selected" => array($segmentId)));
	$actionDuplicateJson['content']['html'] = "New message.";
	
	$newActionDuplicateJson = json_encode($actionDuplicateJson);
	echo $newActionDuplicateJson . "\n";
	
	return ($newActionDuplicateJson);
}

function getActionValidation($segmentId)
{
	//Creation du Json du message pour le test
	$actionValidationData = array(
		'fortest' => true,	//Phase de teste
		'campaignAnalyser' => false,	//Campaign Analyzer : 'true' = oui / 'false' = non
		'testSegments' => array($segmentId),	//Les Ids des differents segments de testes
		'mediaForTest' => null,	//Rediriger tous les tests vers une seule adresse ('null' pour aucune adresse)
		'textandHtml' => true,	//Envoyer la version texte et la version html : 'true' = oui / 'false' = non
		'comments' => null);	//Commentaire ('null' pour aucuns commentaires)
	
	//On affiche le Json
	$actionValidationJson = json_encode($actionValidationData);
	echo $actionValidationJson . "\n";
	
	return ($actionValidationJson);
}

//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////FIN DES CREATION DES JSONS///////////////////////////////////////////////////////////////////////////////////////////////////
/////DEBUT DU PROGRAMME///////////////////////////////////////////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////


//On trouve l'adresse pour la requete
$url = $urlBase . 'targets?unicity=' . $unicity;

//Requete 'GET'
$return = urlGet($url, $xKey);
$result = $return["result"];
$req = $return["req"];

$tab = createTarget($result, $req, $urlBase, $xKey, $unicity);
$req = $tab['req'];
$targetId = $tab['targetId'];

$tab = createSegment($urlBase, $segmentId, $xKey, $req);
$req = $tab['req'];
$segmentId = $tab['segmentId'];

$req = targetToSegment($urlBase, $targetId, $segmentId, $req, $xKey, $unicity);

$tab = duplicateAction($urlBase, $actionId, $req, $xKey);
$req = $tab['req'];
$actionDuplicateId = $tab['actionDuplicateId'];
$actionDuplicateJson = $tab['result'];

$req = updateActionDuplicate($urlBase, $actionDuplicateId, $req, $actionDuplicateJson, $segmentId, $xKey);

$req = testActionDuplicate($urlBase, $actionDuplicateId, $req, $segmentId, $xKey);
	

curl_close($req);


//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////FIN DU PROGRAMME/////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////DEBUT DES FONCTIONS//////////////////////////////////////////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////


function createTarget($result, $req, $urlBase, $xKey, $unicity)
{
	//Verification des reponses
	if ($result == false)
	{
		//La cible n'existe pas, nous devons la creer

		//Affichage de l'erreur
		$info = curl_getinfo($req);
		echo 'Error : ' . $info['http_code'] . " : Creation of the target.\n";

		//Nouvelle url
		$url = $urlBase . 'targets/';

		//On remplit la requete 'POST'
		$con = urlPostOrPut($req, 'POST', getTargetJson($unicity), $xKey, $url);
		echo $con['result'] . "\n";
	}
	else
	{
		//La cible existe, nous la mettons a jour

		//On affiche les anciennes valeurs
		echo $result . "\n";

		//Nouvelle url
		$url = $urlBase . 'targets/';

		//On remplit la requete 'PUT'
		$con = urlPostOrPut($req, 'PUT', getTargetJson($unicity), $xKey, $url);
	}
	$result = $con['result'];
	$info = $con['info'];
	$req = $con['req'];
	
	if ($info['http_code'] != 200)
	{
		echo 'Error : ' . $info['http_code'] . '. End';
		exit (1);
	}
	
	$tab = json_decode($result, TRUE);
	$targetId = $tab['id'];
	
	return (array('req' => $req, 'targetId' => $targetId));
}


function createSegment($urlBase, $segmentId, $xKey, $req)
{
	//On trouve l'addresse pour la requete
	$url = $urlBase . 'segments/' . $segmentId;

	//Connexion pour le segment
	if ($segmentId != null)
		$con = urlPostOrPut($req, 'PUT', getSegmentJson(), $xKey, $url);
	else
		$con = urlPostOrPut($req, 'POST', getSegmentJson(), $xKey, $url);

	$result = $con['result'];
	$info = $con['info'];
	$req = $con['req'];

	if ($info['http_code'] != 200)
	{
		echo 'Error : ' . $info['http_code'] . '. End';
		exit (1);
	}
	
	if ($segmentId == null)
	{
		$tab = json_decode($result, TRUE);
		$segmentId = $tab['id'];
	}
	
	return (array('req' => $req, 'segmentId' => $segmentId));
}

function targetToSegment($urlBase, $targetId, $segmentId, $req, $xKey, $unicity)
{
	echo "Target : " . $targetId . " add to the segment : " . $segmentId . "\n\n";
	
	//Nouvelle url
	$url = $urlBase . 'targets/' . $targetId . '/segments/' . $segmentId;

	//Connexion
	$con = urlPostOrPut($req, 'POST', null, $xKey, $url);

	$result = $con['result'];
	$info = $con['info'];
	$req = $con['req'];

	if ($info['http_code'] != 204)
	{
		echo 'Error : ' . $info['http_code'] . '. Fin';
		exit (1);
	}
	else
		echo 'The target : ' . $unicity . " is add to the segment.\n";
	
	return ($req);
}

function duplicateAction($urlBase, $actionId, $req, $xKey)
{
	//Nouvelle url
	$url = $urlBase . 'actions/' . $actionId . '/duplication';

	//On duplique a l'identique
	$con = urlPostOrPut($req, 'POST', null, $xKey, $url);

	$result = $con['result'];
	$info = $con['info'];
	$req = $con['req'];

	if ($info['http_code'] != 200)
	{
		echo "Error request 'POST' duplication : " . $info['http_code'];
		exit (1);
	}
	else
	{
		//On affiche la cible
		echo "\nThe action duplicate ( " . json_decode($result, TRUE)['name'] . " ) is created.\n";
		echo $result . "\n\n";
	}

	//On recupere l'id de l'action
	$tab = json_decode($result, TRUE);
	$actionDuplicateId = $tab['id'];

	return (array('req' => $req, 'actionDuplicateId' => $actionDuplicateId, 'result' => $result));
}

function updateActionDuplicate($urlBase, $actionDuplicateId, $req, $actionDuplicateJson, $segmentId, $xKey)
{
	//Nouvelle url
	$url = $urlBase . 'actions/' . $actionDuplicateId;
	
	//On modifie l'action
	$con = urlPostOrPut($req, 'PUT', getNewActionDuplicateJson($actionDuplicateJson, $segmentId), $xKey, $url);

	$result = $con['result'];
	$info = $con['info'];
	$req = $con['req'];

	if ($info['http_code'] != 200)
	{
		echo "Error request 'PUT' modification : " . $info['http_code'];
		exit (1);
	}
	else
	{
		echo "The action duplicate ( " . json_decode($result, TRUE)['name'] . " ) is update.\n";
	}

	return ($req);
}

function testActionDuplicate($urlBase, $actionDuplicateId, $req, $segmentId, $xKey)
{
	//On test l'action
	
	//Nouvelle url
	$url = $urlBase . 'actions/' . $actionDuplicateId . '/validation';
	
	echo "\n\n" . $url . "\n\n";
	
	//Connexion
	$con = urlPostOrPut($req, 'POST', getActionValidation($segmentId), $xKey, $url);
	$result = $con['result'];
	$info = $con['info'];
	$req = $con['req'];
	
	//On attend que la phase de test soit terminee (peut prendre plusieurs minutes)
	$actionState = waitForState($actionDuplicateId, $xKey, $urlBase);
	
	//On verifie les reponses
	if ($info['http_code'] != 204 || $actionState == 20)
	{
		if ($actionState == 20)
			echo 'Error : the test failed';
		else
			echo 'Error : ' . $info['http_code'];
	}
	else
	{
		//La phase de test a reussi
		echo "The action is tested.\n\n";
	}
	
	return ($req);	
}

//Utilisation de cURL pour remplir les requetes
//Fonctions de connexion
function startCurlInit($url)
{
	$init = curl_init();
	curl_setopt($init, CURLOPT_URL, $url);
	curl_setopt($init, CURLOPT_RETURNTRANSFER, true);
	return ($init);
}

function urlGet($url, $xKey)
{
	//On remplit la requete 'GET'
	$req = startCurlInit($url);
	curl_setopt($req, CURLOPT_CUSTOMREQUEST, 'GET');

	//Mise en place du xKey et des options
	curl_setopt($req, CURLOPT_HTTPHEADER, array(
	'X-Key: ' . $xKey,
	'Content-Type: application/json'));

	//Execution de la requete
	$result = curl_exec($req);
	return (array('req' => $req, 'result' => $result));
}

function urlPostOrPut($req, $request, $dataJson, $xKey, $url)
{
	//On remplit la requete avec le bon verbe ($request) : POST / PUT
	$req = startCurlInit($url);
	curl_setopt($req, CURLOPT_CUSTOMREQUEST, $request);
	curl_setopt($req, CURLOPT_POSTFIELDS, $dataJson);

	//Mise en place du xKey et des options
	curl_setopt($req, CURLOPT_HTTPHEADER, array(
	'X-Key: ' . $xKey,
	'Content-Type: application/json',
	'Content-Length: ' . strlen($dataJson)));

	//Execution de la requete
	$result = curl_exec($req);
	
	//Verification des reponses
	$info = curl_getinfo($req);
	
	return (array('result' => $result, 'info' => $info, 'req' => $req));
}

//Fonction qui verifie que la phase de test soit fini (peut prendre plusieurs minutes)
function waitForState($idAction, $xKey, $urlBase)
{
	$actionState = 30;

	while ($actionState != 38 && $actionState != 20)
	{
		//On attend 20 secondes
		echo "Wait 20sec...\n";
		sleep(20);

		//On trouve l'adresse pour la requete
		$url = $urlBase . 'actions/' . $idAction;

		//On remplit la requete 'GET'
		$req = startCurlInit($url);
		curl_setopt($req, CURLOPT_CUSTOMREQUEST, 'GET');

		//Mise en place du xKey et des options
		curl_setopt($req, CURLOPT_HTTPHEADER, array(
		'X-Key: ' . $xKey,
		'Content-Type: application/json'));

		//Execution de la requete
		$result = curl_exec($req);

		$tab = json_decode($result, TRUE);
		$actionState = $tab['informations']['state'];
	}
	return ($actionState);
}

?>

File: cookbook/java/duplicateAndAddNewSegment.java
import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;

import org.json.JSONException;
import org.json.JSONObject;

public class duplicateAndAddNewSegment
{
	public static JSONObject getTargetJson(String unicity) throws JSONException
	{
		//Syntaxe pour les differents types d'informations :
		String targetString = "name";	//Chaine de caracteres
		String tagetListOfValues = "Mr";	//Liste de valeurs
		String targetEmail = unicity;	//E-mail
		String tagetPhoneNumber = "0123456789";	//Telephone
		String targetTextZone = "150 caracters max";	//Zone de texte
		int targetNumbers = 123;	//Valeur numerique
		String targetDate = "01/01/2000";	//Date
		String largetListMultipleValues[] = {"valeur 1", "valeur 2"};	//Liste de valeurs multiples

		//Creation du tableau en fonction de l'id des champs de la fiche cible : "id-champ" => "valeur du champ"
		JSONObject targetJson = new JSONObject();
		targetJson.put("5398", targetString);
		targetJson.put("5399", tagetListOfValues);
		targetJson.put("5400", targetEmail);
		targetJson.put("5452", tagetPhoneNumber);
		targetJson.put("5453", targetTextZone);
		targetJson.put("5454", targetNumbers);
		targetJson.put("5455", targetDate);
		targetJson.put("5456", largetListMultipleValues);

		//On affiche le message
		System.out.print("New Json : " + targetJson + "\n");
		
		return (targetJson);
	}
	
	public static JSONObject getSegmentJson() throws JSONException
	{
		//Syntaxe pour les differents types d'informations :
		String segmentType = "static";	//Code pour creer un segment Statique
		String segmentName = "Nom du segment";	//Nom du segment
		String segmentDescription = "Description";	//Description du segment
		String segmentExpiration = "2016-01-08T12:11:00Z";	//Date d'expiration du segment
		Boolean segmentIsTest = true;	//Le test est un segment de test : oui = 'true' / non = 'false'

		//Creation du Json du message
		JSONObject segmentJson = new JSONObject();
		segmentJson.put("type", segmentType);
		segmentJson.put("name", segmentName);
		segmentJson.put("description", segmentDescription);
		segmentJson.put("expiration", segmentExpiration);
		segmentJson.put("isTest", segmentIsTest);

		//On affiche le message
		System.out.print("\nNew Json : " + segmentJson + "\n");
		
		return (segmentJson);
	}
	
	public static JSONObject getNewActionDuplicateJson(JSONObject actionDuplicateJson, String segmentId) throws JSONException
	{
		int[] segmentList = {Integer.parseInt(segmentId)};
		
		//Modification de la nouvelle action
		JSONObject segments = new JSONObject();
		segments.put("selected", segmentList);
		
		JSONObject scheduler = new JSONObject();
		scheduler.put("type", "asap");	//Envoie : immediat = 'asap' / Date = 'scheduled'
		//scheduler.put("type", "2015-07-27T08:15:00Z");	//Si type = 'scheduled' sinon a enlever
		scheduler.put("segments", segments);
		
		actionDuplicateJson.put("scheduler", scheduler);
		
		return (actionDuplicateJson);
	}
	
	public static JSONObject getActionValidation(String segmentId) throws JSONException
	{
		Object NULL = null;
		int[] segmentList = {Integer.parseInt(segmentId)};
		
		JSONObject actionValidationJson = new JSONObject();
		actionValidationJson.put("fortest", true);	//Phase de test
		actionValidationJson.put("campaignAnalyser", false);	//Campaign Analyzer : 'true' = oui / 'false' = non
		actionValidationJson.put("testSegments", segmentList);	//Les Ids des differents segments de tests
		actionValidationJson.put("mediaForTest", NULL);	//Rediriger tous les tests vers une seule adresse ('NULL' pour aucune valeur)
		actionValidationJson.put("textandHtml", false);	//Envoyer la version texte et la version html : 'true' = oui / 'false' = non
		actionValidationJson.put("comments", NULL);	//Commentaire ('NULL' pour aucuns commentaires)
		
		//On affiche le Json
		System.out.print(actionValidationJson + "\n");
		
		return (actionValidationJson);
	}
	
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////FIN DES CREATION DES JSONS//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////DEBUT DU PROGRAMME//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	
	public static void main(String[] args) throws Exception
	{
		//Url de base
		String urlBase = "http://v8.mailperformance.com/";

		//Ici, renseignez l'email de la cible, la X-Key
		String xKey = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
		String unicity = "test@test.com";	//L'adresse mail de la cible a mettre dans le segment ("null" pour ne rien ajouter au segment)
		String segmentId = "0123";	//Id du segment a modifier ("null" si le segment est a creer)
		String actionId = "000ABC";	//Id de l'action a dupliquer
		
		//Ne pas remplir
		String targetId = null;
		ArrayList result = new ArrayList();
		
		
		//Lancement de la connexion pour remplir la requete
		String url = urlBase + "targets?unicity=" + unicity;
		HttpURLConnection con = null;
		
		con = urlGet(url, xKey);
		
		result = createTarget(con, unicity, xKey, urlBase);
		con = (HttpURLConnection) result.get(0);
		targetId = (String) result.get(1);
		
		result = createSegment(segmentId, urlBase, con, xKey);
		con = (HttpURLConnection) result.get(0);
		segmentId = (String) result.get(1);

		con = targetToSegment(urlBase, targetId, segmentId, con, xKey);
		
		result = duplicateAction(urlBase, actionId, con, xKey);
		con = (HttpURLConnection) result.get(0);
		String actionDuplicateId = (String) result.get(1);
		JSONObject actionDuplicateJson = (JSONObject) result.get(2);
		
		con = updateActionDuplicate(urlBase, actionDuplicateId, con, actionDuplicateJson, segmentId, xKey);
		
		con = testActionDuplicate(urlBase, actionDuplicateId, con, xKey, segmentId);
		
		con.disconnect();
		System.exit(1);
	}
	
	
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////FIN DU PROGRAMME////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////DEBUT DES FONCTIONS/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	
	public static ArrayList createTarget(HttpURLConnection con, String unicity, String xKey, String urlBase) throws IOException, JSONException
	{
		String url = urlBase + "targets/";
		
		//Verification des reponses
		int responseCode = con.getResponseCode();
		if (responseCode != 200)
		{
			if (responseCode != 404)
			{
				//Affichage de l'erreur
				System.out.print("Error : " + responseCode + ' ' + con.getResponseMessage());
				System.exit(1);
			}
			else
			{
				System.out.print("Error : " + responseCode + " : Creation of the target.\n");
				con = urlPostOrPut(con, "POST", getTargetJson(unicity), xKey, url);
			}
		}
		else
		{
			//Lecture des donnees
			BufferedReader buffRead = new BufferedReader(new InputStreamReader(con.getInputStream()));
			String reply = buffRead.readLine();
			buffRead.close();
			System.out.print(reply + "\n");
			
			con = urlPostOrPut(con, "PUT", getTargetJson(unicity), xKey, url);
		}
		
		//Lecture des donnees
		BufferedReader buffRead = new BufferedReader(new InputStreamReader(con.getInputStream()));
		String reply = buffRead.readLine();
		buffRead.close();
		System.out.print(reply + "\nData base changed.\n\n");
		
		//On recupere l'id de l'action
		JSONObject jObject  = new JSONObject(reply);
		String targetId = jObject.getString("id");
		
		ArrayList result = new ArrayList();
		result.add(con);
		result.add(targetId);
		
		return (result);
	}
	
	public static ArrayList createSegment(String segmentId, String urlBase, HttpURLConnection con, String xKey) throws MalformedURLException, IOException, JSONException
	{
		//Lancement de la connexion
		if (segmentId != null)
		{
			String url =  urlBase + "segments/" + segmentId;
			con = urlPostOrPut(con, "PUT", getSegmentJson(), xKey, url);
		}
		else
		{
			String url =  urlBase + "segments/";
			con = urlPostOrPut(con, "POST", getSegmentJson(), xKey, url);
		}
		
		//Verification des reponses
		int responseCode = con.getResponseCode();
		if (responseCode != 200)
		{
			//Affichage de l'erreur
			System.out.print("Error : " + responseCode + " " + con.getResponseMessage());
			System.exit(1);
		}
		
		//Lecture des donnees ligne par lignes
		BufferedReader buffRead = new BufferedReader(new InputStreamReader(con.getInputStream()));
		String reply = buffRead.readLine();
		con.disconnect();
		buffRead.close();
	
		//On recupere l'id de l'action
		JSONObject jObject  = new JSONObject(reply);
		String segmentName = jObject.getString("name");
		segmentId = Integer.toString(jObject.getInt("id"));
		
		
		//Le segment a bien ete cree
		System.out.print("The segment static : " + segmentName + " is created / update.\n\n");
		
		ArrayList result = new ArrayList();
		result.add(con);
		result.add(segmentId);
		
		return (result);
	}
	
	public static HttpURLConnection targetToSegment(String urlBase, String targetId, String segmentId, HttpURLConnection con, String xKey) throws MalformedURLException, IOException
	{
		//Nouvelle url
		String url = urlBase + "targets/" + targetId + "/segments/" + segmentId;
		
		//Connexion
		con = urlPostOrPut(con, "POST", new JSONObject(), xKey, url);
		
		System.out.print("The target (" + targetId + ") is add to the segment (" + segmentId + ").\n\n");
		
		return (con);
	}
	
	public static ArrayList duplicateAction(String urlBase, String actionId, HttpURLConnection con, String xKey) throws MalformedURLException, IOException, JSONException
	{
		//Nouvelle url
		String url = urlBase + "actions/" + actionId + "/duplication";
		
		//On duplique a l'identique
		con = urlPostOrPut(con, "POST", new JSONObject(), xKey, url);
						
		//Lecture des donnees
		BufferedReader buffRead = new BufferedReader(new InputStreamReader(con.getInputStream()));
		String reply = buffRead.readLine();
		con.disconnect();
		buffRead.close();
		
		//On affiche la cible
		System.out.print("The action duplicate ( " + new JSONObject(reply).getString("name") + " ) is created :\n");
		System.out.print(reply + "\n\n");
		
		//On recupere l'id de la nouvelle action
		JSONObject actionDuplicateJson  = new JSONObject(reply);
		String actionDuplicateId = actionDuplicateJson.getString("id");
		
		ArrayList result = new ArrayList();
		result.add(con);
		result.add(actionDuplicateId);
		result.add(actionDuplicateJson);
		
		return (result);
		
	}
	
	public static HttpURLConnection updateActionDuplicate(String urlBase, String actionDuplicateId, HttpURLConnection con, JSONObject actionDuplicateJson, String segmentId, String xKey) throws MalformedURLException, IOException, JSONException
	{
		//Nouvelle url
		String url = urlBase + "actions/" + actionDuplicateId;
			
		//On duplique a l'identique
		con = urlPostOrPut(con, "PUT", getNewActionDuplicateJson(actionDuplicateJson, segmentId), xKey, url);

		//Lecture des donnees
		BufferedReader buffRead = new BufferedReader(new InputStreamReader(con.getInputStream()));
		String reply = buffRead.readLine();
		con.disconnect();
		buffRead.close();
		
		System.out.print("The action duplicate ( " + new JSONObject(reply).getString("name") + " ) is update:\n");
		System.out.print(reply + "\n\n");
		
		return (con);
	}
	
	public static HttpURLConnection testActionDuplicate(String urlBase, String actionDuplicateId, HttpURLConnection con, String xKey, String segmentId) throws MalformedURLException, InterruptedException, IOException, JSONException
	{
		//Nouvelle url
		String url = urlBase + "actions/" + actionDuplicateId + "/validation";
		
		//Lancement de la connexion
		con = urlPostOrPut(con, "POST", getActionValidation(segmentId), xKey, url);
		
		int actionState = waitForState(actionDuplicateId, xKey, urlBase);
		
		if (actionState == 20)
		{
			//Affichage de l'erreur
			System.out.print("Error : the test failed.");
		}
		else
		{
			//La phase de test a reussi
			System.out.print("The action is tested.\n\n");
		}
		return (con);
	}
	
	//Fonctions de connexion
	public static HttpURLConnection startCon(String url, String xKey) throws MalformedURLException, IOException
	{
		//Lancement de la connexion
		HttpURLConnection con = (HttpURLConnection)new URL(url).openConnection();
		
		//Mise en place du xKey et des options
		con.setRequestProperty("X-Key", xKey);
		con.setRequestProperty("Content-Type", "application/json");
		return (con);
	}

	public static HttpURLConnection urlGet(String url, String xKey) throws MalformedURLException, IOException
	{
		//On remplit la requete 'GET'
		HttpURLConnection con = startCon(url, xKey);
		con.setRequestMethod("GET");
		
		return (con);
	}
	
	public static HttpURLConnection urlPostOrPut(HttpURLConnection con, String request, JSONObject dataJson, String xKey, String url) throws MalformedURLException, IOException
	{
		con.disconnect();
		//Lancement de la connexion pour remplir la requete
		con = startCon(url, xKey);
		con.setRequestProperty("Content-Length", Integer.toString(dataJson.toString().length()));
		con.setRequestMethod(request);
		con.setDoOutput(true);
		
		// Envoie des informations dans la connexion
	    DataOutputStream streamCon = new DataOutputStream(con.getOutputStream());
	    streamCon.write(dataJson.toString().getBytes());
	    streamCon.flush();
	    streamCon.close();
	 
		//Verification des reponses
		int responseCode = con.getResponseCode();
		if (responseCode != 200 && responseCode != 204)
		{
			//Affichage de l'erreur
			System.out.print("Error : " + responseCode + ' ' + con.getResponseMessage() + ".\nEnd urlPostOrPut");
			System.exit(1);
		}
		return (con);
	}
	
	//Fonction d'attente de fin de test
	public static int waitForState(String idAction, String xKey, String urlBase) throws InterruptedException, MalformedURLException, IOException, JSONException
	{
		int actionState = 30;
		
		while (actionState != 38 && actionState != 20)
		{
			//On attend 20 secondes
			System.out.print("Wait 20sec...\n");
			Thread.sleep(20000);
			
			//Nouvelle adresse
			String url = urlBase + "actions/" + idAction;
			
			//Lancement de la connexion pour remplir la requete
			HttpURLConnection con = urlGet(url, xKey);
			
			//Lecture des donnees ligne par lignes
			BufferedReader buffRead = new BufferedReader(new InputStreamReader(con.getInputStream()));
			String reply = buffRead.readLine();
			con.disconnect();
			buffRead.close();
			
			//On recupere l'etat de l'action
			JSONObject jObject  = new JSONObject(reply);
			actionState = jObject.getJSONObject("informations").getInt("state");
		}
		return (actionState);
	}
}
